File: low-level/sorts/sorts.py
arr = [25, 47, 92, 70, 64, 89, 25, 13, 31, 6, 58];

def insertion_sort(array):      # in-place sort
    for i in xrange(1, len(array)):
        j = i
        while j and array[j] < array[j - 1]:
            array[j], array[j - 1] = array[j - 1], array[j]
            j -= 1

def merge_sort(array):          # recursive out-of-place sort
    length = len(array)
    def merge(left, right):
        result = []
        while left and right:
            if left[0] < right[0]:
                result.append(left.pop(0))
            else:
                result.append(right.pop(0))
        result.extend(left if left else right)
        return result

    if length <= 1:
        return array
    left, right = merge_sort(array[:(length / 2)]), merge_sort(array[(length / 2):])
    return merge(left, right)

File: low-level/sorts/sorts.c
#include <stdio.h>

void insertion_sort(int *array, int length) {   // in-place sort
    int i, j, value;
    for (i = 1; i < length; i++) {
        j = i - 1;
        value = array[i];
        while (j >= 0 && array[j] > value) {
            array[j + 1] = array[j];    // overwrites the values in front and moves forward
            j -= 1;
        } array[j + 1] = value;
    }
}

void main() {
    int arr[] = {25, 47, 92, 70, 64, 89, 25, 13, 31, 6, 58};
    int length = sizeof(arr) / sizeof(arr[0]);
    insertion_sort(arr, length);
}
